Repository-Wide Tracing Migration & Full Network Layer Instrumentation

sntntn opened the pull request:
### Description

This PR completes the **repository-wide migration** from `logging` to `tracing` and provides **full tracing instrumentation of the network layer**. 
> This PR covers **Phase 2 (Network Instrumentation)** as well as the final log to tracing migration step from **Phase 4 (Final migration)**
[logging-to-tracing roadmap](https://github.com/eth-protocol-fellows/cohort-six/blob/master/projects/Grandine-Implementing-Tokio-Tracing-For-Debugging-And-Performance-Analysis.md#roadmap)

The `eth2_libp2p` submodule has been updated to directly use tracing (in [**PR**](https://github.com/sntntn/eth2_libp2p/pull/3)), removing the previously used slog-stdlog bridge. This change eliminates the logging overhead and simplifies the logging flow (detailed in [issue](https://github.com/sntntn/grandine/issues/8)), ensuring consistent and richer logging features (like structured fields and spans) directly within eth2_libp2p, allowing logs—including PEER_LOG_METRICS—to be easily parsed and processed by monitoring or analysis tools


🎯 **PR Goal**
- **Global migration**
     - Introduce tracing globally across the entire project.
     - Remove `slog` usage from `eth2_libp2p` & `p2p` crate
     - Coordinate the migration of tracing in the `eth2_libp2p` **submodule** ([**PR**](https://github.com/sntntn/eth2_libp2p/pull/3)) with the **main** repo.   
 
- **Network Layer Instrumentation**
     - Integrate tracing features (spans + structured fields) in the network layer. 
        - Add tracing spans for `eth2_libp2p` instrumentation
        - Instrument key http_api endpoints and internal submission functions:
           - Block publishing (`publish_block`, `publish_blinded_block`, `publish_signed_block`)
           - Validator requests (`validator_block`, `validator_blinded_block`, `validator_attestation_data`) 
            - Attestation submissions (`submit_pool_attestations`, `submit_attestations_to_pool`)
            - Server run functions: `run` and `run_internal`

- **Helper Macros**
     - Implement `PEER_LOG_METRICS`-aware helper macros globally, including submodule 
     - Implement the `crit!` macro to log **critical errors**:



- **issues**:
    - #3  

    - #4 

    - #7 

    - #8 

    - #10 
    
    - #11
    
    - #12 
    
    - #13 

    - #14 

    - #15 




### Notes on instrumentation:
- All spans for API and critical internal functions are marked as root-level (parent=None) to ensure independent tracing, avoiding inheritance from unrelated spans.
- Fields like block_root, slot, and attestation_count are included for structured logging and observability.
- `level = trace` can be configured for detailed logs in development, while production logging can use higher levels to reduce verbosity.

---

🧪 **How to Test**

To ensure that the migration to tracing works correctly, follow these steps:


1. Clone the grandine repository:
```
git clone https://github.com/sntntn/grandine.git
```

2. Switch to the feature/network-instrumentation branch in the grandine repository:
```
cd grandine
git checkout feature/network-instrumentation
```

3. Sync and update the eth2_libp2p submodule:
```
git submodule sync eth2_libp2p
git submodule update --init --recursive --remote eth2_libp2p
```

4. Checkout the tracing branch inside the eth2_libp2p submodule:
```
cd eth2_libp2p
git checkout tracing
cd ..
```

5. Build and run the project with the desired features enabled:
*To test the logging at a specific level, you can set the GRANDINE_LOG environment variable. For example, to set the debug level for `eth2_libp2p`:*
```
GRANDINE_LOG=info cargo run --profile compact --features default-networks,logger-always-write-style --bin grandine --   --network holesky   --checkpoint-sync-url https://checkpoint-sync.holesky.ethpandaops.io
```

sntntn committed:
http_api: Add tracing instrumentation to main tasks and critical endpoints

- Added `tracing::instrument` spans to the main HTTP task and internal server startup logic
- Instrumented key public-facing endpoints such as block publishing, validator block requests, and attestation submissions
- Added internal spans for important block and attestation processing functions

Relates to sntntn/grandine#3

sntntn committed:
Introduce tracing and replace log macros with tracing PEER_LOG_METRICS-aware helpers (partial)

Relates to sntntn/grandine#7

sntntn committed:
Introduce tracing and replace log macros with tracing PEER_LOG_METRICS-aware helpers (part 2 of multiple commits)

Relates to sntntn/grandine#7

sntntn committed:
Introduce tracing and replace log macros with tracing PEER_LOG_METRICS-aware helpers (part 3 of multiple commits)

Relates to sntntn/grandine#7

sntntn committed:
Introduces the trace_with_peers! macro and applies it to relevant tracing calls.

Relates to sntntn/grandine#7 and sntntn/grandine#6

sntntn committed:
Introduce tracing and replace log macros with tracing PEER_LOG_METRICS-aware helpers (P2P part of multiple commits)

Relates to stntn/grandine#7 and sntntn/grandine#4

sntntn committed:
Point eth2_libp2p submodule to fork repository

Relates to sntntn/grandine#8

sntntn committed:
Add crit! macro to support eth2_libp2p backport

Relates to #8

sntntn committed:
sync with eth2_libp2p submodule (remove unused argument)

Relates to #8

sntntn committed:
Eliminate slog bridge(p2p) & sync with eth2_libp2p submodule

Relates to sntntn/grandine#8

sntntn committed:
Upgrade _with_peers macros to support structured tracing logs

Relates to #10 & #7

sntntn committed:
make _with_peers and crit! macros fully compatible with tracing-style argument passing

Relates to #10.

sntntn committed:
Make http_api spans root-level & sync with eth2_libp2p submodule

parent=None for independent API operations
Relates to #3

Tumas commented:
[**issue**](https://conventionalcomments.org/): Seems like `EnvFilter` in `binary_utils/src/lib.rs` does not apply to dependencies that use tracing (discv5) or log (refinery). 

<img width="1446" height="203" alt="Screenshot from 2025-08-28 16-01-49" src="https://github.com/user-attachments/assets/8523cd8c-12f6-4525-8176-d00e4bcf3f72" />
<img width="1337" height="136" alt="refinery" src="https://github.com/user-attachments/assets/97880e76-8e0d-4bab-8c19-246c5c8df2f3" />

This is different from existing setup, where filter in `binary_utils/src/lib.rs` is global, and allows us to control dependencies as well.

Tumas commented:
[**note**](https://conventionalcomments.org/): there are multiple linter errors that should be fixed or exceptions should be added. You can check them by running `./scripts/ci/clippy.bash` script.

Also check the tests too: ` cargo test --release `
And please run `cargo fmt`

sntntn commented:
> [**issue**](https://conventionalcomments.org/): Seems like `EnvFilter` in `binary_utils/src/lib.rs` does not apply to dependencies that use tracing (discv5) or log (refinery).
> 
> <img alt="Screenshot from 2025-08-28 16-01-49" width="1446" height="203" src="https://private-user-images.githubusercontent.com/88451/483162044-8523cd8c-12f6-4525-8176-d00e4bcf3f72.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTYzOTE3ODMsIm5iZiI6MTc1NjM5MTQ4MywicGF0aCI6Ii84ODQ1MS80ODMxNjIwNDQtODUyM2NkOGMtMTJmNi00NTI1LTgxNzYtZDAwZTRiY2YzZjcyLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA4MjglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwODI4VDE0MzEyM1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWYwYjQ3YThlNGQ5MDBiYTIyNjZkZWYyNDllYzYzNTJlYTA4ZGIxOTAyMjEzZDNjNjJjODBlMzQ0Mjg4ZDZjZTEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.sEzXLucSMNhN2PsiIl8tHHK5i3aQYNqB3TwgW76EdBE"> <img alt="refinery" width="1337" height="136" src="https://private-user-images.githubusercontent.com/88451/483162047-97880e76-8e0d-4bab-8c19-246c5c8df2f3.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTYzOTE3ODMsIm5iZiI6MTc1NjM5MTQ4MywicGF0aCI6Ii84ODQ1MS80ODMxNjIwNDctOTc4ODBlNzYtOGUwZC00YmFiLThjMTktMjQ2YzVjOGRmMmYzLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA4MjglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwODI4VDE0MzEyM1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTQ1YjY2N2E2NzU5YjgzMDEyMDkzMzRkNDJjYjUxYzhhNDIzODQzNGI2ODAwOWViZTM1OGM4NmE1YWZjODc2MGEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.MNL4bUO48b6jVDUioDqmyN06Pe6u_ASTy3KmPoN3sbA">
> This is different from existing setup, where filter in `binary_utils/src/lib.rs` is global, and allows us to control dependencies as well.

Thanks for the feedback!
Just to clarify - when you mention that the EnvFilter does not apply to dependencies (like discv5 or refinery), do you mean that setting something like
```
GRANDINE_LOG=discv5=debug
```
or 
```
GRANDINE_LOG=refinery_core=debug

```

I tried these cases locally and the logs did respect the filter, so I’d like to understand which exact scenario you had in mind in order to reproduce it.
<img width="1261" height="249" alt="Screenshot from 2025-08-28 16-52-49" src="https://github.com/user-attachments/assets/24dfdf0b-3fd7-49ee-9af8-208e1b992d40" />


<img width="1265" height="195" alt="Screenshot from 2025-08-28 17-02-02" src="https://github.com/user-attachments/assets/3004295e-4183-4eb2-b349-9d0bbe709f55" />

Tumas commented:
I meant that `.add_directive(LevelFilter::OFF.into())` is not a global switch-off for dependencies.

We should not see output from dependencies unless they are whitelisted. `discv5` for example, emits lots of warnings which obscure normal log. `refinery` is not mentioned in `grandine_utils/src/lib.rs` but we do see the output from it.

sntntn commented:
> I meant that `.add_directive(LevelFilter::OFF.into())` is not a global switch-off for dependencies.
> 
> We should not see output from dependencies unless they are whitelisted. `discv5` for example, emits lots of warnings which obscure normal log. `refinery` is not mentioned in `grandine_utils/src/lib.rs` but we do see the output from it. 
> 
> 

Now it's completely clear, it will be done

sntntn committed:
eth2_libp2p sync + clippy cleanup: inline format strings

Relates to #12

sntntn committed:
cargo fmt

Relates to #12

sntntn committed:
refactor(features): replace log/warn functions with macros for testable logging

Relates to #13

sntntn committed:
set cognitive-complexity-threshold

as debugging macros with peers seem to inflate counted complexity
Relates to #12

sntntn committed:
Rename test file for clarity

Relates to #13

sntntn committed:
fix(feature): add logging and tracing deps for feature macros

Relates to #13

sntntn committed:
cleanup clippy lints

Relates to #12

sntntn committed:
add clippy expectation for complex type

Relates to #12

sntntn committed:
eth2_libp2p sync + cargo fmt

Relates to #12

sntntn committed:
remove expect clippy comments

sntntn commented:
Everything is ready for review again.

- `cargo test --release` -> all tests passed
- `./scripts/ci/clippy.bash` -> no errors or warnings reported
- Applied `cargo fmt` 

pending @Tumas

Tumas commented on ad_hoc_bench/Cargo.toml:
```diff
@@ -18,7 +18,8 @@ fork_choice_control = { workspace = true }
 futures = { workspace = true }
 itertools = { workspace = true }
-tracing ={ workspace = true }
+tracing = { workspace = true }
```
[**chore**](https://conventionalcomments.org/): Keep dependencies in alphabetical order

Tumas commented on Cargo.toml:
```diff
@@ -440,6 +436,7 @@ tempfile = '3'
 test-case = '3'
 test-generator = '0.3'
 testing_logger = '0.1'
+tracing-test = "0.2"
```
[**chore**](https://conventionalcomments.org/): Keep dependencies in alphabetical order

Tumas commented on attestation_verifier/Cargo.toml:
```diff
@@ -17,7 +17,8 @@ fork_choice_store = { workspace = true }
 helper_functions = { workspace = true }
 itertools = { workspace = true }
-log = { workspace = true }
+tracing = { workspace = true }
```
[**chore**](https://conventionalcomments.org/): Keep dependencies in alphabetical order

Tumas commented on block_producer/Cargo.toml:
```diff
@@ -24,7 +24,8 @@ helper_functions = { workspace = true }
 keymanager = { workspace = true }
-log = { workspace = true }
+tracing = { workspace = true }
+logging ={ workspace = true }
```
[**chore**](https://conventionalcomments.org/): Keep dependencies in alphabetical order

Tumas commented on builder_api/Cargo.toml:
```diff
@@ -17,7 +17,8 @@ helper_functions = { workspace = true }
 http_api_utils = { workspace = true }
 itertools = { workspace = true }
-log = { workspace = true }
+tracing = { workspace = true }
```
[**chore**](https://conventionalcomments.org/): Keep dependencies in alphabetical order

Tumas commented on clippy.toml:
```diff
@@ -1,4 +1,7 @@
 avoid-breaking-exported-api = false
+
```
[**nitpick**](https://conventionalcomments.org/): add a comment documenting why this was needed. This will prevent people from making false assumptions and eliminates an extra git-blame step.

sntntn replied:
makes sense, I will

Tumas commented on data_dumper/Cargo.toml:
```diff
@@ -8,7 +8,8 @@ anyhow = { workspace = true }
 fs-err = { workspace = true }
-log = { workspace = true }
+tracing = { workspace = true }
+logging = { workspace = true }
```
[**chore**](https://conventionalcomments.org/): Keep dependencies in alphabetical order

Tumas commented on database/Cargo.toml:
```diff
@@ -14,7 +14,8 @@ futures = { workspace = true }
 libmdbx = { workspace = true }
-log = { workspace = true }
+tracing ={ workspace = true }
+logging = { workspace = true }
```
[**chore**](https://conventionalcomments.org/): Keep dependencies in alphabetical order

Tumas commented on debug_info/Cargo.toml:
```diff
@@ -12,6 +12,8 @@ bytesize = { workspace = true }
 futures = { workspace = true }
 log = { workspace = true }
+tracing = { workspace = true }
+logging ={ workspace = true }
```
[**chore**](https://conventionalcomments.org/): Keep dependencies in alphabetical order

Tumas commented on deposit_tree/Cargo.toml:
```diff
@@ -14,6 +14,8 @@ ssz = { workspace = true }
 typenum = { workspace = true }
 types = { workspace = true }
+tracing = { workspace = true }
+logging = { workspace = true }
```
[**chore**](https://conventionalcomments.org/): Keep dependencies in alphabetical order

Tumas commented on doppelganger_protection/Cargo.toml:
```diff
@@ -10,7 +10,8 @@ bls = { workspace = true }
 liveness_tracker = { workspace = true }
-log = { workspace = true }
+tracing ={ workspace = true }
+logging = { workspace = true }
```
[**chore**](https://conventionalcomments.org/): Keep dependencies in alphabetical order

Tumas commented on eth1/Cargo.toml:
```diff
@@ -17,7 +17,8 @@ fs-err = { workspace = true }
 itertools = { workspace = true }
-log = { workspace = true }
+tracing ={ workspace = true }
+logging = { workspace = true }
```
[**chore**](https://conventionalcomments.org/): Keep dependencies in alphabetical order

Tumas commented on eth1_api/Cargo.toml:
```diff
@@ -28,7 +28,8 @@ hex = { workspace = true }
 itertools = { workspace = true }
 jwt-simple = { workspace = true }
-log = { workspace = true }
+logging = { workspace = true }
```
[**chore**](https://conventionalcomments.org/): Keep dependencies in alphabetical order

Tumas commented on execution_engine/Cargo.toml:
```diff
@@ -14,7 +14,8 @@ eth2_libp2p = { workspace = true }
 hex = { workspace = true }
-log = { workspace = true }
+logging = { workspace = true }
+tracing = { workspace = true }
```
[**chore**](https://conventionalcomments.org/): Keep dependencies in alphabetical order

Tumas commented on features/Cargo.toml:
```diff
@@ -8,9 +8,12 @@ workspace = true
 [dependencies]
 log = { workspace = true }
+tracing = { workspace = true }
+logging = { workspace = true }
```
[**chore**](https://conventionalcomments.org/): Keep dependencies in alphabetical order

Tumas commented on fork_choice_control/Cargo.toml:
```diff
@@ -26,7 +26,8 @@ genesis = { workspace = true }
 http_api_utils = { workspace = true }
 itertools = { workspace = true }
-log = { workspace = true }
+logging = { workspace = true }
```
[**chore**](https://conventionalcomments.org/): Keep dependencies in alphabetical order

Tumas commented on fork_choice_store/Cargo.toml:
```diff
@@ -22,7 +22,8 @@ helper_functions = { workspace = true }
 kzg_utils = { workspace = true }
-log = { workspace = true }
+tracing = { workspace = true }
+logging = { workspace = true }
```
[**chore**](https://conventionalcomments.org/): Keep dependencies in alphabetical order

Tumas commented on http_api/Cargo.toml:
```diff
@@ -27,7 +27,8 @@ helper_functions = { workspace = true }
 itertools = { workspace = true }
 liveness_tracker = { workspace = true }
-log = { workspace = true }
+tracing = { workspace = true }
```
[**chore**](https://conventionalcomments.org/): Keep dependencies in alphabetical order

Tumas commented on http_api/src/standard.rs:
```diff
@@ -2267,6 +2269,7 @@ pub async fn validator_aggregate_attestation_v2<P: Preset, W: Wait>(
 }
 
 /// `GET /eth/v1/validator/blinded_blocks/{slot}`
+#[instrument( parent = None, skip_all, fields(slot = %slot))]
```
[**question**](https://conventionalcomments.org/): Is there a reason for the extra space after `(`? It seems to be used inconsistently.

sntntn replied:
no reason, I'll push fix for typo space

Tumas commented on keymanager/Cargo.toml:
```diff
@@ -15,7 +15,8 @@ fs-err = { workspace = true }
 itertools = { workspace = true }
-log = { workspace = true }
+logging = { workspace = true }
+tracing = { workspace = true }
```
[**chore**](https://conventionalcomments.org/): Keep dependencies in alphabetical order

Tumas commented on liveness_tracker/Cargo.toml:
```diff
@@ -14,7 +14,8 @@ fork_choice_control = { workspace = true }
 itertools = { workspace = true }
-log = { workspace = true }
+logging = { workspace = true }
+tracing = { workspace = true }
```
[**chore**](https://conventionalcomments.org/): Keep dependencies in alphabetical order

Tumas commented on metrics/Cargo.toml:
```diff
@@ -19,7 +19,8 @@ futures = { workspace = true }
 http_api_utils = { workspace = true }
-log = { workspace = true }
+logging = { workspace = true }
+tracing = { workspace = true }
```
[**chore**](https://conventionalcomments.org/): Keep dependencies in alphabetical order

Tumas commented on operation_pools/Cargo.toml:
```diff
@@ -23,7 +23,8 @@ good_lp = { workspace = true }
 itertools = { workspace = true }
-log = { workspace = true }
+tracing = { workspace = true }
+logging = { workspace = true }
```
[**chore**](https://conventionalcomments.org/): Keep dependencies in alphabetical order

Tumas commented on p2p/Cargo.toml:
```diff
@@ -28,6 +28,7 @@ helper_functions = { workspace = true }
 igd-next = { workspace = true }
 itertools = { workspace = true }
 log = { workspace = true }
+tracing ={ workspace = true }
```
[**chore**](https://conventionalcomments.org/): Keep dependencies in alphabetical order

Tumas commented on prometheus_metrics/Cargo.toml:
```diff
@@ -8,7 +8,8 @@ authors = ["Grandine <info@grandine.io>"]
 features = { workspace = true }
 futures = { workspace = true }
-log = { workspace = true }
+tracing = { workspace = true }
```
[**chore**](https://conventionalcomments.org/): Keep dependencies in alphabetical order

Tumas commented on pubkey_cache/Cargo.toml:
```diff
@@ -11,7 +11,8 @@ anyhow = { workspace = true }
 dashmap = { workspace = true }
 database = { workspace = true }
-log = { workspace = true }
+logging = { workspace = true }
```
[**chore**](https://conventionalcomments.org/): Keep dependencies in alphabetical order

Tumas commented on runtime/Cargo.toml:
```diff
@@ -32,7 +32,8 @@ http_api = { workspace = true }
 liveness_tracker = { workspace = true }
-log = { workspace = true }
+tracing = { workspace = true }
+logging = { workspace = true }
```
[**chore**](https://conventionalcomments.org/): Keep dependencies in alphabetical order

Tumas commented on signer/Cargo.toml:
```diff
@@ -16,7 +16,8 @@ doppelganger_protection = { workspace = true }
 helper_functions = { workspace = true }
 itertools = { workspace = true }
-log = { workspace = true }
+tracing = { workspace = true }
```
[**chore**](https://conventionalcomments.org/): Keep dependencies in alphabetical order

Tumas commented on slasher/Cargo.toml:
```diff
@@ -17,7 +17,8 @@ eth1_api = { workspace = true }
 futures = { workspace = true }
 helper_functions = { workspace = true }
-log = { workspace = true }
+tracing = { workspace = true }
```
[**chore**](https://conventionalcomments.org/): Keep dependencies in alphabetical order

Tumas commented on slashing_protection/Cargo.toml:
```diff
@@ -15,7 +15,8 @@ fs-err = { workspace = true }
 itertools = { workspace = true }
-log = { workspace = true }
+tracing = { workspace = true }
+logging = { workspace = true }
```
[**chore**](https://conventionalcomments.org/): Keep dependencies in alphabetical order

Tumas commented on state_cache/Cargo.toml:
```diff
@@ -10,7 +10,8 @@ workspace = true
 anyhow = { workspace = true }
 im = { workspace = true }
-log = { workspace = true }
+tracing = { workspace = true }
```
[**chore**](https://conventionalcomments.org/): Keep dependencies in alphabetical order

Tumas commented on validator/Cargo.toml:
```diff
@@ -32,7 +32,8 @@ http_api_utils = { workspace = true }
 keymanager = { workspace = true }
 liveness_tracker = { workspace = true }
-log = { workspace = true }
+tracing = { workspace = true }
```
[**chore**](https://conventionalcomments.org/): Keep dependencies in alphabetical order

Tumas commented on validator_key_cache/Cargo.toml:
```diff
@@ -9,7 +9,8 @@ anyhow = { workspace = true }
 eip_2335 = { workspace = true }
 fs-err = { workspace = true }
-log = { workspace = true }
+tracing = { workspace = true }
```
[**chore**](https://conventionalcomments.org/): Keep dependencies in alphabetical order

Tumas commented on validator_statistics/Cargo.toml:
```diff
@@ -11,6 +11,8 @@ anyhow = { workspace = true }
 fork_choice_control = { workspace = true }
 helper_functions = { workspace = true }
+tracing = { workspace = true }
+logging = { workspace = true }
```
[**chore**](https://conventionalcomments.org/): Keep dependencies in alphabetical order

sntntn committed:
typo extra space

sntntn committed:
chore: sort dependencies alphabetically

sntntn committed:
docs(clippy): explain higher cognitive-complexity threshold
